rb_link/src/server/getter.rs

```diff
@@ -0,0 +1,190 @@
+use super::{B2RMessage, B2RServer};
+use std::collections::{HashMap, VecDeque};
+use std::sync::{Arc, Mutex};
+
+/// A getter to get message from the bluesim by id
+pub struct IDGetter {
+    b2r_cache: Arc<Mutex<HashMap<u32, VecDeque<B2RMessage>>>>,
+}
+
+impl IDGetter {
+    /// Crate a new getter bind to the given server
+    pub fn new(server: &B2RServer) -> Self {
+        IDGetter {
+            b2r_cache: server.b2r_cache.clone(),
+        }
+    }
+    /// Return the earliest message from the probe with id.
+    /// This function will block until there is a message available for retrieval.
+    pub fn get(&mut self, id: u32) -> B2RMessage {
+        loop {
+            let mut b2r_cache = self.b2r_cache.lock().expect("Fail to lock b2r_cache");
+            if let Some(queue) = b2r_cache.get_mut(&id) {
+                if let Some(b2r_message) = queue.pop_front() {
+                    return b2r_message;
+                }
+            }
+            drop(b2r_cache);
+        }
+    }
+
+    /// Get all message send by the probe with id.
+    pub fn get_id_all(&mut self, id: u32) -> Vec<B2RMessage> {
+        let mut b2r_cache = self.b2r_cache.lock().expect("Fail to lock b2r_cache");
+        let mut messages: Vec<B2RMessage> = Vec::new();
+
+        if let Some(queue) = b2r_cache.get_mut(&id) {
+            while let Some(b2r_message) = queue.pop_front() {
+                messages.push(b2r_message);
+            }
+        }
+        messages
+    }
+
+    /// Return the earliest message from the probe with id.
+    /// This function will return None if there is no message available for retrieval.
+    pub fn try_get(&mut self, id: u32) -> Option<B2RMessage> {
+        let mut b2r_cache = self.b2r_cache.lock().expect("Fail to lock b2r_cache");
+        if let Some(queue) = b2r_cache.get_mut(&id) {
+            if let Some(b2r_message) = queue.pop_front() {
+                return Some(b2r_message);
+            }
+        }
+        None
+    }
+}
+
+/// A getter that retrieves messages from bluesim sequentially according to cycles.
+pub struct CycleGetter {
+    b2r_cache: Arc<Mutex<HashMap<u32, VecDeque<B2RMessage>>>>,
+}
+
+impl CycleGetter {
+    /// Crate a new getter bind to the given server
+    pub fn new(server: &B2RServer) -> Self {
+        CycleGetter {
+            b2r_cache: server.b2r_cache.clone(),
+        }
+    }
+
+    /// Get all the messages sent by the earliest cycle.
+    /// If there are no messages available, it will return an empty Vec.
+    pub fn get_cycle_message(&mut self) -> Vec<B2RMessage> {
```

Comment:
naming problem. Why the important "earliest" not show in the method name but show in the comment?